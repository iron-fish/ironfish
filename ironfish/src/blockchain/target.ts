/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import type Serde from '../serde'

function max(a: BigInt, b: BigInt): BigInt {
  if (a > b) {
    return a
  } else {
    return b
  }
}

/**
 * Courtesy of https://coolaj86.com/articles/convert-js-bigints-to-typedarrays/
 *
 * Convert a Buffer to a big integer number, in big endian format.
 *
 * I'm concerned about efficiency here. Converting a string and back and... WTF?
 * Every block hash attempt has to be converted to a Target, so this is a function
 * that should be optimized. We may want to compile this to wasm if there isn't
 * a less janky way to do it.
 *
 * I'm pushing it out like this for now so I can focus on bigger architecture concerns.
 *
 * Sorry.
 */
export function bytesToBigInt(bytes: Buffer): bigint {
  const hex: string[] = []
  if (bytes.length === 0) {
    return BigInt(0)
  }
  bytes.forEach(function (i) {
    let h = i.toString(16)
    if (h.length % 2) {
      h = '0' + h
    }
    hex.push(h)
  })

  return BigInt('0x' + hex.join(''))
}

export function bigIntToBytes(bigint: bigint): Buffer {
  let hex = bigint.toString(16)
  if (hex.length % 2) {
    hex = '0' + hex
  }

  const len = hex.length / 2
  const u8 = Buffer.alloc(len)

  let i = 0
  let j = 0
  while (i < len) {
    u8[i] = parseInt(hex.slice(j, j + 2), 16)
    i += 1
    j += 2
  }

  return u8
}

/**
 * The bound divisor of the difficulty, used to update difficulty (and subsequently target).
 * We are taking in large part Ethereum's dynamic difficulty calculation,
 * with the exeption of 'uncles' and 'difficulty bomb' as a concept
 * https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.md
 * original algorithm:
 * diff = (parent_diff +
 *         (parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) // 10, -99))
 *        ) + 2**((current_block_number // 100000) — 2)
 * Note we are not including the difficulty bomb (which is this part: 2**((current_block_number // 100000) — 2))
 * So the algorithm for target is:
 * diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
 * note that timestamps above are in seconds, and JS timestamps are in ms
 * The bound divisor of the difficulty is the '2048' part of that equation
 */
const DIFFICULTY_ADJUSTMENT_DENOMINATOR = 2048

/**
 *  Minimum difficulty, which is equivalent to maximum target
 */
const MIN_DIFFICULTY = BigInt(131072)

/**
 * Maximum target, which is equivalent of minimum difficulty of 131072
 * target == 2**256 / difficulty
 */
const MAX_TARGET = BigInt(
  '883423532389192164791648750371459257913741948437809479060803100646309888',
)

/**
 *  Maximum number to represent a 256 bit number, which is 2**256 - 1
 */
const MAX_256_BIT_NUM = BigInt(
  '115792089237316195423570985008687907853269984665640564039457584007913129639935',
)

export class Target {
  targetValue: bigint
  constructor(targetValue: bigint | BigInt | Buffer | string | number | undefined = undefined) {
    if (targetValue === undefined) {
      this.targetValue = BigInt(0)
    } else {
      let candidate
      if (targetValue instanceof Buffer) {
        candidate = bytesToBigInt(targetValue)
      } else {
        candidate = BigInt(targetValue)
      }
      if (candidate > MAX_256_BIT_NUM) {
        throw new Error('Target value exceeds max target')
      } else {
        this.targetValue = candidate
      }
    }
  }

  /**
   * Maximum target (in terms of difficulty), which is equivalent of
   * minimum difficulty of 131072
   * maximum target == minimum difficulty
   * target == 2**256 / difficulty
   */
  static maxTarget(): Target {
    return new Target(MAX_TARGET)
  }

  /**
   * The initial target on the genesis block.
   *
   * This will need to be hand-tuned.
   */
  static initialTarget(): Target {
    return this.maxTarget()
  }

  /**
   * Calculate the target for the current block given the timestamp in that
   * block's header, the pervious block's timestamp and previous block's target.
   *
   * To verify whether a target on a block is correct, pass in the timestamp in its header,
   * its previous block's timestamp, and its previous block's target
   * and compare the resulting target to what is specified on the current block header
   *
   * @param time the block's timestamp for which the target is calculated for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateTarget(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockTarget: Target,
  ): Target {
    const parentDifficulty = previousBlockTarget.toDifficulty()
    const difficulty = Target.calculateDifficulty(
      time,
      previousBlockTimestamp,
      parentDifficulty,
    )

    return Target.fromDifficulty(difficulty)
  }

  /**
   *
   * Calculate the difficulty for the current block given the timestamp in that
   * block's header, the pervious block's timestamp and previous block's target.
   *
   * Note that difficulty == 2**256 / target and target == 2**256 / difficulty
   *
   * Returns the difficulty for a block given it timestamp for that block and its parent.
   * @param time the block's timestamp for which the target is calcualted for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateDifficulty(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockDifficulty: BigInt,
  ): BigInt {
    // We are taking in large part Ethereum's dynamic difficulty calculation,
    // with the exeption of 'uncles' and 'difficulty bomb' as a concept
    // https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.md
    // original algorithm:
    // diff = (parent_diff +
    //         (parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) // 10, -99))
    //        ) + 2**((current_block_number // 100000) — 2)
    // Note we are not including the difficulty bomb (which is this part: 2**((current_block_number // 100000) — 2))
    // So the algorithm we're taking is:
    // diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    // note that timestamps above are in seconds, and JS timestamps are in ms

    // max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    const diffInSeconds = (time.getTime() - previousBlockTimestamp.getTime()) / 1000
    const sign = BigInt(Math.max(1 - Math.floor(diffInSeconds / 10), -99))
    const offset = BigInt(previousBlockDifficulty) / BigInt(DIFFICULTY_ADJUSTMENT_DENOMINATOR)

    // diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    const difficulty = max(
      BigInt(previousBlockDifficulty) + offset * sign,
      Target.minDifficulty(),
    )

    return difficulty
  }

  /**
   * Returns the minimum difficulty that can be used for Iron Fish blocks
   * To be used in calculateTarget for easier mocking
   */
  static minDifficulty(): BigInt {
    return MIN_DIFFICULTY
  }

  /**
   * Converts difficulty to Target
   */
  static fromDifficulty(difficulty: BigInt): Target {
    if (difficulty == BigInt(1)) {
      return new Target(MAX_256_BIT_NUM)
    }
    return new Target(BigInt(2) ** BigInt(256) / BigInt(difficulty))
  }

  /**
   * Return the difficulty representation as a big integer
   */
  toDifficulty(): bigint {
    return BigInt(2) ** BigInt(256) / this.targetValue
  }

  /**
   * Add the given amout to the target's value. A negative amount makes the target
   * harder to achieve, a positive one makes it easier.
   *
   * If adjustment would make target negative or higher than max allowed value,
   * the current target is returned unchanged.
   */
  adjust(amount: bigint): Target {
    let adjusted = this.targetValue + amount
    if (adjusted > MAX_TARGET || adjusted < 0) {
      adjusted = this.targetValue
    }
    return new Target(adjusted)
  }

  /**
   * Return whether or not this target meets the requirements of the given target,
   * which is to say, this has a lower numeric value then the provided one.
   */
  static meets(hashValue: BigInt, target: Target): boolean {
    return hashValue <= target.targetValue
  }

  /**
   * Return the target number as a big integer
   */
  asBigInt(): bigint {
    return this.targetValue
  }

  /**
   * Big endian, and since even after 20 years in the industry, I can't remember
   * what that means, the most significant byte is in the 0th index of the array.
   *
   * The resulting byte array is always left padded with 0s to be 32 bytes long
   */
  asBytes(): Buffer {
    const bytes = bigIntToBytes(this.targetValue)
    const result = Buffer.alloc(32)
    result.set(bytes, 32 - bytes.length)
    return result
  }
}

export class TargetSerde implements Serde<Target, string> {
  equals(target1: Target, target2: Target): boolean {
    return target1.targetValue === target2.targetValue
  }

  serialize(target: Target): string {
    return target.targetValue.toString()
  }

  deserialize(data: string): Target {
    if (typeof data !== 'string') {
      throw new Error('Can only deserialize Target from string')
    }
    return new Target(data)
  }
}

export const TargetSerdeInstance = new TargetSerde()
