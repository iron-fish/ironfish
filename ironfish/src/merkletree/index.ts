/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

export { MerkleTree, Side as WitnessSide } from './merkletree'
export { MerkleHasher, ConcatHasher, StructureHasher, RangeHasher, NoteHasher } from './hasher'
export { Witness } from './witness'
