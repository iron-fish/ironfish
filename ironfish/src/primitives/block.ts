/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { zip } from 'lodash'
import { Assert } from '../assert'
import { BlockHeader, BlockHeaderSerde, SerializedBlockHeader } from './blockheader'
import { MintDescription } from './mintDescription'
import { NoteEncrypted, NoteEncryptedHash } from './noteEncrypted'
import { Nullifier } from './nullifier'
import { SerializedTransaction, Transaction } from './transaction'

/**
 * The hash used in the "previousHash" field on the initial block in the
 * chain. The initial block is intentionally invalid, so we need to special
 * case it.
 */
export const GENESIS_BLOCK_PREVIOUS = Buffer.alloc(32)

export const GENESIS_BLOCK_SEQUENCE = 1

export const GRAFFITI_SIZE = 32

/**
 * Represent a single block in the chain. Essentially just a block header
 * and the list of transactions that were added to the tree between the
 * previous block and the ones committed to in this header.
 */
export class Block {
  header: BlockHeader
  transactions: Transaction[]

  constructor(header: BlockHeader, transactions: Transaction[]) {
    this.header = header
    this.transactions = transactions
  }

  /**
   * Get the number of notes and nullifiers stored on this block.
   */
  counts(): SerializedCounts {
    let notes = 0
    let nullifiers = 0

    for (const transaction of this.transactions) {
      notes += transaction.notes.length
      nullifiers += transaction.spends.length
    }

    return { notes, nullifiers }
  }

  /**
   * Get a list of all spends on transactions in this block.
   *
   * Note: there is no spend on a miner's fee transaction in the header
   */
  *spends(): Generator<{
    nullifier: Nullifier
    commitment: NoteEncryptedHash
    size: number
  }> {
    for (const transaction of this.transactions) {
      for (const spend of transaction.spends) {
        yield spend
      }
    }
  }

  /**
   * Get a list of all notes created in this block including the miner's fee
   * note on the header.
   */
  *notes(): Generator<NoteEncrypted> {
    for (const transaction of this.transactions) {
      for (const note of transaction.notes) {
        yield note
      }
    }
  }

  /**
   * Get a list of all mints on transactions in this block.
   */
  *mints(): Generator<MintDescription> {
    for (const transaction of this.transactions) {
      for (const mint of transaction.mints) {
        yield mint
      }
    }
  }

  equals(block: Block): boolean {
    if (block === this) {
      return true
    }

    if (!this.header.equals(block.header)) {
      return false
    }

    if (this.transactions.length !== block.transactions.length) {
      return false
    }

    for (const [transaction1, transaction2] of zip(this.transactions, block.transactions)) {
      if (!transaction1 || !transaction2 || !transaction1.equals(transaction2)) {
        return false
      }
    }

    return true
  }

  get minersFee(): Transaction {
    const tx = this.transactions[0]
    Assert.isNotUndefined(tx, 'Block has no miners fee')
    return tx
  }

  toCompactBlock(): CompactBlock {
    const header = this.header

    const [minersFee, ...transactions] = this.transactions
    const transactionHashes = transactions.map((t) => t.hash())

    return {
      header,
      transactionHashes,
      transactions: [
        {
          index: 0,
          transaction: minersFee,
        },
      ],
    }
  }
}

export type CompactBlockTransaction = {
  index: number
  transaction: Transaction
}

export type CompactBlock = {
  header: BlockHeader
  transactionHashes: Buffer[]
  transactions: CompactBlockTransaction[]
}

export type SerializedBlock = {
  header: SerializedBlockHeader
  transactions: SerializedTransaction[]
}

export type SerializedCounts = { notes: number; nullifiers: number }

export class BlockSerde {
  static serialize(block: Block): SerializedBlock {
    return {
      header: BlockHeaderSerde.serialize(block.header),
      transactions: block.transactions.map((t) => t.serialize()),
    }
  }

  static deserialize(data: SerializedBlock): Block {
    if (
      typeof data === 'object' &&
      data !== null &&
      'header' in data &&
      'transactions' in data &&
      Array.isArray(data.transactions)
    ) {
      const header = BlockHeaderSerde.deserialize(data.header)
      const transactions = data.transactions.map((t) => new Transaction(t))
      return new Block(header, transactions)
    }
    throw new Error('Unable to deserialize')
  }
}
