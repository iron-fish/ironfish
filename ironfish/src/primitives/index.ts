/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

export { Block } from './block'
export { BlockHeader } from './blockheader'
export { Spend } from './spend'
export { Target } from './target'
export { Transaction } from './transaction'
