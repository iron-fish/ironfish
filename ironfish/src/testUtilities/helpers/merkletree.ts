/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import bufio from 'bufio'
import { MerkleTree } from '../../merkletree'
import { NodeValue } from '../../merkletree/database/nodes'
import { StructureHasher } from '../../merkletree/hasher'
import { Side } from '../../merkletree/merkletree'
import { IDatabase, IDatabaseEncoding, StringEncoding } from '../../storage'
import { createTestDB } from '../helpers/storage'

type StructureLeafValue = {
  merkleHash: string
  parentIndex: number
}

class StructureLeafEncoding implements IDatabaseEncoding<StructureLeafValue> {
  serialize(value: StructureLeafValue): Buffer {
    const bw = bufio.write()

    bw.writeVarString(value.merkleHash, 'utf8')
    bw.writeU32(value.parentIndex)

    return bw.render()
  }

  deserialize(buffer: Buffer): StructureLeafValue {
    const bw = bufio.read(buffer, true)

    const merkleHash = bw.readVarString('utf8')
    const parentIndex = bw.readU32()

    return {
      merkleHash,
      parentIndex,
    }
  }
}

class StructureNodeEncoding implements IDatabaseEncoding<NodeValue<string>> {
  serialize(value: NodeValue<string>): Buffer {
    const bw = bufio.write()

    bw.writeVarString(value.hashOfSibling, 'utf8')

    if (value.side === Side.Left) {
      bw.writeU8(0)
      bw.writeU32(value.parentIndex)
    } else {
      bw.writeU8(1)
      bw.writeU32(value.leftIndex)
    }

    return bw.render()
  }

  deserialize(buffer: Buffer): NodeValue<string> {
    const reader = bufio.read(buffer, true)

    const hashOfSibling = reader.readVarString('utf8')

    const sideNumber = reader.readU8()
    const side = sideNumber === 0 ? Side.Left : Side.Right

    const otherIndex = reader.readU32()

    if (side === Side.Left) {
      const leftNode = {
        side,
        hashOfSibling,
        parentIndex: otherIndex,
      } as const
      return leftNode
    }

    const rightNode = {
      side,
      hashOfSibling,
      leftIndex: otherIndex,
    } as const
    return rightNode
  }
}

export async function makeTree({
  name,
  db,
  location,
  depth,
  leaves,
}: {
  name?: string
  db?: IDatabase
  location?: string
  depth?: number
  leaves?: string
} = {}): Promise<MerkleTree<string, string, string, string>> {
  if (!db) {
    const { db: database } = await createTestDB(undefined, location)
    db = database
  }

  const tree = new MerkleTree({
    hasher: new StructureHasher(),
    leafIndexKeyEncoding: new StringEncoding(),
    leafEncoding: new StructureLeafEncoding(),
    nodeEncoding: new StructureNodeEncoding(),
    db: db,
    name: name,
    depth: depth,
    defaultValue: '~',
  })

  await db.open()

  if (leaves) {
    await tree.addBatch(leaves)
  }

  return tree
}
