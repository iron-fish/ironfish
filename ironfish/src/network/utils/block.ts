/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import bufio from 'bufio'
import { Assert } from '../../assert'
import { SerializedBlock } from '../../primitives/block'
import { GraffitiSerdeInstance } from '../../serde/serdeInstances'
import { BigIntUtils } from '../../utils/bigint'

export function writeBlock(
  bw: bufio.StaticWriter | bufio.BufferWriter,
  block: SerializedBlock,
): bufio.StaticWriter | bufio.BufferWriter {
  const { header, transactions } = block

  bw.writeU32(header.sequence)
  bw.writeHash(header.previousBlockHash)
  bw.writeHash(header.noteCommitment.commitment)
  bw.writeU32(header.noteCommitment.size)
  bw.writeHash(header.nullifierCommitment.commitment)
  bw.writeU32(header.nullifierCommitment.size)
  bw.writeBytes(BigIntUtils.toBytesLE(BigInt(header.target), 32))
  bw.writeBytes(BigIntUtils.toBytesLE(BigInt(header.randomness), 8))
  bw.writeU64(header.timestamp)

  Assert.isTrue(BigInt(header.minersFee) <= 0)
  bw.writeBytes(BigIntUtils.toBytesLE(-BigInt(header.minersFee), 8))

  bw.writeBytes(GraffitiSerdeInstance.deserialize(header.graffiti))

  bw.writeU16(transactions.length)
  for (const transaction of transactions) {
    bw.writeVarBytes(transaction)
  }

  return bw
}

export function readBlock(reader: bufio.BufferReader): SerializedBlock {
  const sequence = reader.readU32()
  const previousBlockHash = reader.readHash('hex')
  const noteCommitment = reader.readHash()
  const noteCommitmentSize = reader.readU32()
  const nullifierCommitment = reader.readHash('hex')
  const nullifierCommitmentSize = reader.readU32()
  const target = BigIntUtils.fromBytesLE(reader.readBytes(32)).toString()
  const randomness = BigIntUtils.fromBytesLE(reader.readBytes(8)).toString()
  const timestamp = reader.readU64()
  const minersFee = (-BigIntUtils.fromBytesLE(reader.readBytes(8))).toString()
  const graffiti = GraffitiSerdeInstance.serialize(reader.readBytes(32))

  const transactionsLength = reader.readU16()
  const transactions = []
  for (let j = 0; j < transactionsLength; j++) {
    transactions.push(reader.readVarBytes())
  }

  return {
    header: {
      sequence,
      previousBlockHash,
      noteCommitment: {
        commitment: noteCommitment,
        size: noteCommitmentSize,
      },
      nullifierCommitment: {
        commitment: nullifierCommitment,
        size: nullifierCommitmentSize,
      },
      target,
      randomness,
      timestamp,
      minersFee,
      graffiti,
    },
    transactions,
  }
}

export function getBlockSize(block: SerializedBlock): number {
  let size = 0
  size += 4 // sequence
  size += 32 // previousBlockHash
  size += 32 // noteCommitment.commitment
  size += 4 // noteCommitment.size
  size += 32 // nullifierCommitment.commitment
  size += 4 // nullifierCommitment.size
  size += 32 // target
  size += 8 // randomness
  size += 8 // timestamp
  size += 8 // minersFee
  size += 32 // graffiti

  size += 2 // transactions length
  for (const transaction of block.transactions) {
    size += bufio.sizeVarBytes(transaction)
  }
  return size
}
