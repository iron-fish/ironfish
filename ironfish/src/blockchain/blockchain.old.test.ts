/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { RangeHasher } from '../merkletree'
import { Block } from '../primitives/block'
import { Target } from '../primitives/target'
import {
  addNotes,
  blockHash,
  makeChainInitial,
  makeFakeBlock,
  makeNullifier,
  SerializedTestTransaction,
  TestBlockchain,
  TestStrategy,
  TestTransaction,
} from '../testUtilities/fake'

describe('Note adding', () => {
  const strategy = new TestStrategy(new RangeHasher())
  let blockchain: TestBlockchain
  let listener: jest.Mock
  let targetSpy: jest.SpyInstance
  beforeEach(async () => {
    targetSpy = jest.spyOn(Target, 'minDifficulty').mockReturnValue(BigInt(1))
    blockchain = await makeChainInitial(strategy)
    listener = jest.fn()
    blockchain.onHeadChange.on(listener)
  })

  afterAll(() => [targetSpy.mockClear()])

  it('immediately adds in order notes to the tree', async () => {
    await blockchain.addNote(0, 'zero')
    await blockchain.addNote(1, 'one')
    expect(await blockchain.notes.size()).toBe(2)
    expect(await blockchain.notes.get(0)).toBe('zero')
    expect(await blockchain.notes.get(1)).toBe('one')
    expect(await blockchain.nullifiers.size()).toBe(0)
    expect(listener).not.toBeCalled()
  })

  it("throws if the note doesn't match the previously inserted note that position", async () => {
    await blockchain.addNote(0, 'zero')
    await blockchain.addNote(1, 'one')
    await blockchain.addNote(2, 'two')
    await expect(blockchain.addNote(2, 'not two')).rejects.toThrowError(
      `Tried to insert a note, but a different note already there for position 2`,
    )
  })

  it('throws if the position is larger than the number of notes', async () => {
    await blockchain.addNote(0, 'zero')
    await blockchain.addNote(1, 'one')
    await blockchain.addNote(2, 'two')
    await expect(blockchain.addNote(4, 'not two')).rejects.toThrowError(
      `Can't insert a note at index 4. Merkle tree has a count of 3`,
    )
  })
})

describe('Nullifier adding', () => {
  const strategy = new TestStrategy(new RangeHasher())
  let blockchain: TestBlockchain
  let listener: jest.Mock
  beforeEach(async () => {
    blockchain = await makeChainInitial(strategy)
    listener = jest.fn()
    blockchain.onHeadChange.on(listener)
  })

  it('immediately adds in order nullifiers to the tree', async () => {
    const nullifier1 = Buffer.alloc(32)
    const nullifier2 = makeNullifier(1)
    await blockchain.addNullifier(0, nullifier1)
    await blockchain.addNullifier(1, nullifier2)
    expect(await blockchain.nullifiers.size()).toBe(2)
    expect(await blockchain.nullifiers.get(0)).toEqualNullifier(nullifier1)
    expect(await blockchain.nullifiers.get(1)).toEqualNullifier(nullifier2)
    expect(await blockchain.notes.size()).toBe(0)
    expect(listener).not.toBeCalled()
  })

  it("throws if the nullifier doesn't match the previously inserted nullifier in that position", async () => {
    const nullifier0 = Buffer.alloc(32)
    const nullifier1 = makeNullifier(1)
    const nullifier2 = makeNullifier(2)
    await blockchain.addNullifier(0, nullifier0)
    await blockchain.addNullifier(1, nullifier1)
    await blockchain.addNullifier(2, nullifier2)
    await expect(blockchain.addNullifier(2, nullifier0)).rejects.toThrowError(
      `Tried to insert a nullifier, but a different nullifier already there for position 2`,
    )
  })

  it('throws if the position is larger than the number of nullifiers', async () => {
    await blockchain.addNullifier(0, makeNullifier(0))
    await blockchain.addNullifier(1, makeNullifier(1))
    await blockchain.addNullifier(2, makeNullifier(2))
    await expect(blockchain.addNullifier(4, makeNullifier(4))).rejects.toThrowError(
      `Can't insert a nullifier at index 4. Merkle tree has a count of 3`,
    )
  })

  it('sixNullifierRoot matches expected rootHash', async () => {
    await blockchain.addNullifier(0, makeNullifier(0))
    await blockchain.addNullifier(1, makeNullifier(1))
    await blockchain.addNullifier(2, makeNullifier(2))
    await blockchain.addNullifier(3, makeNullifier(3))
    await blockchain.addNullifier(4, makeNullifier(4))
    await blockchain.addNullifier(5, makeNullifier(5))
    const rootHash = await blockchain.nullifiers.rootHash()
    expect(rootHash.equals(sixNullifierRoot)).toBeTruthy()
  })
})

const sixNullifierRoot = Buffer.from([
  225,
  164,
  205,
  91,
  37,
  68,
  206,
  32,
  128,
  69,
  41,
  50,
  240,
  78,
  211,
  128,
  227,
  49,
  167,
  139,
  132,
  31,
  71,
  88,
  44,
  71,
  19,
  28,
  204,
  126,
  14,
  152,
])

describe('Block matches', () => {
  const strategy = new TestStrategy(new RangeHasher())

  it('is true for block that passes all checks', async () => {
    const blockchain = await makeChainInitial(strategy)
    const header = makeFakeBlock(strategy, blockHash(1), blockHash(2), 2, 3, 5).header
    await addNotes(blockchain, [1, 2, 3, 4, 5])
    await blockchain.nullifiers.add(Buffer.alloc(32))
    header.nullifierCommitment.commitment = await blockchain.nullifiers.rootHash()
    expect((await blockchain.verifier.blockMatchesTrees(header)).valid).toBe(true)
  })
  it("is false if there aren't enough notes in the tree", async () => {
    const anchor = await makeChainInitial(strategy)
    const header = makeFakeBlock(strategy, blockHash(1), blockHash(2), 2, 3, 5).header
    await addNotes(anchor, [1, 2, 3, 4])
    await anchor.nullifiers.add(Buffer.alloc(32))
    header.nullifierCommitment.commitment = await anchor.nullifiers.rootHash()
    expect((await anchor.verifier.blockMatchesTrees(header)).valid).toBe(false)
  })
  it("is false if there aren't enough nullifiers in the tree", async () => {
    const anchor = await makeChainInitial(strategy)
    const header = makeFakeBlock(strategy, blockHash(1), blockHash(2), 2, 3, 5).header
    await addNotes(anchor, [1, 2, 3, 4, 5])
    await anchor.nullifiers.add(Buffer.alloc(32))
    const secondNullifier = Buffer.alloc(32)
    secondNullifier[0] = 1
    await anchor.nullifiers.add(secondNullifier)
    header.nullifierCommitment.commitment = await anchor.nullifiers.rootHash()
    header.nullifierCommitment.size = 8
    expect((await anchor.verifier.blockMatchesTrees(header)).valid).toBe(false)
  })
  it('is false if the note hash is incorrect', async () => {
    const anchor = await makeChainInitial(strategy)
    const header = makeFakeBlock(strategy, blockHash(1), blockHash(2), 2, 3, 5).header

    await addNotes(anchor, [1, 2, 3, 4, 5])
    await anchor.nullifiers.add(Buffer.alloc(32))
    header.nullifierCommitment.commitment = await anchor.nullifiers.rootHash()
    header.noteCommitment.commitment = 'NOOO'
    expect((await anchor.verifier.blockMatchesTrees(header)).valid).toBe(false)
  })
  it('is false for block that has incorrect nullifier hash', async () => {
    const anchor = await makeChainInitial(strategy)
    const header = makeFakeBlock(strategy, blockHash(1), blockHash(2), 2, 3, 5).header
    await addNotes(anchor, [1, 2, 3, 4, 5])
    await anchor.nullifiers.add(Buffer.alloc(32))
    expect((await anchor.verifier.blockMatchesTrees(header)).valid).toBe(false)
  })
})

describe('New block', () => {
  const strategy = new TestStrategy(new RangeHasher())
  let blockchain: TestBlockchain
  let listener: jest.Mock
  let targetSpy: jest.SpyInstance
  let targetMeetsSpy: jest.SpyInstance

  beforeEach(async () => {
    targetSpy = jest.spyOn(Target, 'minDifficulty').mockImplementation(() => BigInt(1))
    targetMeetsSpy = jest.spyOn(Target, 'meets').mockImplementation(() => true)
    blockchain = await makeChainInitial(strategy)
    listener = jest.fn()
    blockchain.onHeadChange.on(listener)
  })

  afterAll(() => {
    jest.useRealTimers()
    targetSpy.mockClear()
    targetMeetsSpy.mockClear()
  })

  it('throws an error if the provided transactions are invalid', async () => {
    await blockchain.nullifiers.add(Buffer.alloc(32))
    const block1 = makeFakeBlock(strategy, blockHash(0), blockHash(1), 1, 1, 2)
    block1.transactions[0]._spends.push({
      nullifier: Buffer.alloc(32),
      commitment: '1-1',
      size: 1,
    })
    block1.header.nullifierCommitment.commitment = await blockchain.nullifiers.rootHash()
    const block2 = makeFakeBlock(strategy, blockHash(1), blockHash(2), 2, 3, 5)
    await blockchain.addBlock(block1)
    await blockchain.addBlock(block2)
    const fakeBlock = makeFakeBlock(strategy, blockHash(0), blockHash(0), 1, 9, 14)
    fakeBlock.transactions[0]._spends.push({
      nullifier: Buffer.alloc(32),
      commitment: '1-1',
      size: 2,
    })
    const minersFee = new TestTransaction(true, ['1'], 1)
    await expect(
      blockchain.newBlock(fakeBlock.transactions, minersFee),
    ).rejects.toMatchInlineSnapshot(`[Error: Miner's fee is incorrect]`)
  })
})

describe('Header consistency is valid against previous', () => {
  let strategy: TestStrategy
  let blockchain: TestBlockchain
  let block2: Block<string, string, TestTransaction, string, string, SerializedTestTransaction>
  let block3: Block<string, string, TestTransaction, string, string, SerializedTestTransaction>

  beforeEach(async () => {
    strategy = new TestStrategy(new RangeHasher())
    blockchain = await makeChainInitial(strategy)

    block2 = makeFakeBlock(strategy, blockHash(1), blockHash(2), 2, 6, 9)
    block3 = makeFakeBlock(strategy, blockHash(2), blockHash(3), 3, 10, 13)
  })

  it("is invalid when the note commitments aren't the same size", async () => {
    block2.header.noteCommitment.size = 99
    await blockchain.addBlock(block2)
    await blockchain.addBlock(block3)

    expect(blockchain.verifier.isValidAgainstPrevious(block3, block3.header))
      .toMatchInlineSnapshot(`
        Object {
          "reason": "Note commitment sizes do not match",
          "valid": false,
        }
      `)
  })

  it("is invalid when the nullifier commitments aren't the same size", async () => {
    block2.header.nullifierCommitment.size = 99
    await blockchain.addBlock(block2)
    await blockchain.addBlock(block3)

    expect(blockchain.verifier.isValidAgainstPrevious(block3, block3.header))
      .toMatchInlineSnapshot(`
      Object {
        "reason": "Note commitment sizes do not match",
        "valid": false,
      }
    `)
  })

  it('Is invalid when the timestamp is in past', async () => {
    await blockchain.addBlock(block2)
    await blockchain.addBlock(block3)
    block3.header.timestamp = new Date(100)

    expect(blockchain.verifier.isValidAgainstPrevious(block3, block3.header))
      .toMatchInlineSnapshot(`
      Object {
        "reason": "Note commitment sizes do not match",
        "valid": false,
      }
    `)
  })

  it('Is invalid when the sequence is wrong', async () => {
    await blockchain.addBlock(block2)
    await blockchain.addBlock(block3)
    block3.header.sequence = 99

    expect(blockchain.verifier.isValidAgainstPrevious(block3, block3.header))
      .toMatchInlineSnapshot(`
      Object {
        "reason": "Note commitment sizes do not match",
        "valid": false,
      }
    `)
  })

  it('is valid when it is valid', async () => {
    await blockchain.addBlock(block2)
    await blockchain.addBlock(block3)

    expect(blockchain.verifier.isValidAgainstPrevious(block3, block3.header))
      .toMatchInlineSnapshot(`
      Object {
        "reason": "Note commitment sizes do not match",
        "valid": false,
      }
    `)
  })
})
