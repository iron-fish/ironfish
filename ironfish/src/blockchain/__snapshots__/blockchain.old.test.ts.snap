// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note adding logs errors if the note doesn't match the previously inserted note that position 1`] = `[Error: Tried to insert a note, but a different note already there for position 2]`;

exports[`Nullifier adding warns if the note doesn't match the previously inserted note that position 1`] = `[Error: Tried to insert a nullifier, but a different nullifier already there for position 2]`;
